### PR TITLE
Add options for `jwks-rsa` to enable cache and rate limiting

### PR DESCRIPTION
**Pull request by cdwills**
Add options to take advantage of the built in caching from `jwks-rsa`.  Description is here: https://github.com/auth0/node-jwks-rsa#caching  

To see the caching in action add `DEBUG=jwks` to your local env for the previous version vs this one.  You'll see the caching configured and in action.

image

Bonus:
- bumped `eslint`, `nyc`, `nock`
- bumped `boom` 
- added option to run Travis tests in node 10.
* I believe we can remove support for `node 6, 7` in the future.

### CODE REVIEW

File: .travis.yml [hunk 1]
```diff
@@ -1,7 +1,7 @@
 language: node_js
 node_js:
+  - '10'
   - '8'
-  - '7'
   - '6'
```
Comment (dbertram) [inline .travis.yml:5]: Why did you remove support for 7, but keep 6 around? B/c evens = LTS?

File: index.js [hunk 2]
```diff
@@ -29,15 +30,19 @@ const decode = partialRight(jwt.decode, [{ complete: true }])
...
-  Promise.reject(Boom.wrap(err, 401))
+  Promise.reject(Boom.unauthorized(err))
 
-const factory = opts => {
+const jwksOptsDefaults = { jwks: { cache: true, rateLimit: true } }
+
+const factory = options => {
   const clients    = {}
-  const verifyOpts = dissoc('issWhitelist', opts)
+  const opts = merge(options, jwksOptsDefaults)
```
Comment (dbertram) [inline index.js]: Should the order of the args to this `merge` be switched?

As-is, won't the values in `jwksOptsDefaults` trump what's passed in via `options`?

Comment (dbertram) [reply index.js]: Follow-up: worth writing a test for?

**Review by dbertram — APPROVED** [review]
One possible bug around merging of jwks option defaults, but otherwise lgtm

File: .travis.yml (hunk 1 above)
Comment (cdwills) [reply .travis.yml:5]: Yes.  Even though 6 will have to go soon too.  Many of the modules we use are cutting support for older versions pretty ruthlessly.

File: index.js (hunk 2 above)
Comment (cdwills) [reply index.js]: 👀

**Comment by cdwills** [comment]
@dbertram great catch on the merge, i changed it up.  Tested locally and feeling good about this.  Would like some final 👀

File: index.js [hunk 3]
```diff
@@ -68,3 +73,5 @@ const factory = opts => {
 }
 
 module.exports = factory
+
+module.exports = factory
```
Comment (dbertram) [inline index.js]: Double export?

**Review by dbertram — APPROVED** [review]
Looks like one piece of :broom:ing left and this lgtm

File: index.js (hunk 3 above)
Comment (cdwills) [reply index.js]: just to be sure, obviously. 🤦‍♂️

**Comment by dbertram** [comment]
image

**Comment by pklingem** [comment]
:eyes:

File: README.md [hunk 4]
```diff
@@ -39,3 +37,25 @@ const handler = req =>
   authentic(req.cookies.token)
     .then(/* the JWT has been validated */)
 ```
+
+## Options
+
+`authentic` accepts a JSON object of options that will be passed to the underlying libraries responsibile for validation.
```
Comment (pklingem) [inline README.md]: the options section is a bit confusing.  Am I correct that any options other that `jwks` or `issWhitelist` will be passed to `jwt.verify()`?  If so, should it be:

```json
{
  verify: { ... },
  jwks: { ... },
  issWhitelist: ...,
}
```

for some futureproofing?  It seems odd to separate `jwks`' options, but not `verify`'s and I could imagine that causing conflicts eventually.

**Review by pklingem — COMMENTED** [review]
just the one comment, other than that, I can confirm that I'm seeing caching on my side.  First request shows the keys, subsequent requests do not.

**Comment by dbertram** [comment]
@pklingem @cdwills Looks like the CI builds are failing due to reduced code coverage:

> ERROR: Coverage for branches (80%) does not meet global threshold (100%)
> ----------|----------|----------|----------|----------|-------------------|
> File      |  % Stmts | % Branch |  % Funcs |  % Lines | Uncovered Line #s |
> ----------|----------|----------|----------|----------|-------------------|
> All files |      100 |       80 |      100 |      100 |                   |
>  index.js |      100 |       80 |      100 |      100 |             45,46 |
> ----------|----------|----------|----------|----------|-------------------|
> error Command failed with exit code 1.
> info Visit https://yarnpkg.com/en/docs/cli/run for documentation about this command.
> error Command failed with exit code 1.
> info Visit https://yarnpkg.com/en/docs/cli/run for documentation about this command.
> The command "yarn run test:ci" exited with 1.

:sadpanda:

**Comment by cdwills** [comment]
💯

**Review by dbertram — APPROVED** [review]

**Comment by dbertram** [comment]
image
